File: backend/app/routes/me.py
import uuid

from flask import Blueprint, jsonify, request
from flask_jwt_extended import get_jwt_identity, jwt_required

from app.constants import (
    MAX_AUTHORIZED_PICKUPS,
    MAX_DELIVERY_ADDRESSES,
    PICKUP_ID_TYPE_LABELS,
    PICKUP_ID_TYPES,
)
from app.models.user import User
from app.services.authorized_pickup_service import (
    create_authorized_pickup,
    delete_authorized_pickup,
    get_authorized_pickup,
    list_authorized_pickups,
    update_authorized_pickup,
)
from app.services.delivery_address_service import (
    create_delivery_address,
    delete_delivery_address,
    get_delivery_address,
    list_delivery_addresses,
    set_default_delivery_address,
    update_delivery_address,
)
from app.services.profile_service import change_password, update_profile
from app.services.warehouse_service import build_shipping_address

me_bp = Blueprint("me", __name__)


def _get_current_user():
    user_id = get_jwt_identity()
    try:
        uid = uuid.UUID(user_id)
    except (TypeError, ValueError):
        return None
    return User.query.get(uid)


@me_bp.route("/me", methods=["GET"])
@jwt_required()
def get_me():
    user = _get_current_user()
    if not user:
        return jsonify({"error": "User not found"}), 404

    return jsonify(
        {
            "user": user.to_dict(
                include_trn=user.role == "customer",
                include_clerk_fields=user.role in ("clerk", "admin"),
            )
        }
    )


@me_bp.route("/me", methods=["PATCH"])
@jwt_required()
def update_me():
    user = _get_current_user()
    if not user:
        return jsonify({"error": "User not found"}), 404

    data = request.get_json(silent=True) or {}
    try:
        user = update_profile(user, data)
    except ValueError as exc:
        return jsonify({"error": str(exc)}), 400

    return jsonify({"user": user.to_dict(include_trn=True)})


@me_bp.route("/me/change-password", methods=["POST"])
@jwt_required()
def change_my_password():
    user = _get_current_user()
    if not user:
        return jsonify({"error": "User not found"}), 404

    data = request.get_json(silent=True) or {}
    try:
        change_password(
            user,
            data.get("current_password") or "",
            data.get("new_password") or "",
        )
    except ValueError as exc:
        return jsonify({"error": str(exc)}), 400

    return jsonify({"message": "Password updated"})


@me_bp.route("/me/shipping-address", methods=["GET"])
@jwt_required()
def get_shipping_address():
    user = _get_current_user()
    if not user:
        return jsonify({"error": "User not found"}), 404

    return jsonify(
        {
            "shipping_id": user.shipping_id,
            "shipping_address": build_shipping_address(user.shipping_id),
        }
    )


@me_bp.route("/me/delivery-addresses", methods=["GET"])
@jwt_required()
def get_delivery_addresses():
    user = _get_current_user()
    if not user:
        return jsonify({"error": "User not found"}), 404

    addresses = list_delivery_addresses(user)
    return jsonify(
        {
            "addresses": [a.to_dict() for a in addresses],
            "max_addresses": MAX_DELIVERY_ADDRESSES,
        }
    )


@me_bp.route("/me/delivery-addresses", methods=["POST"])
@jwt_required()
def create_my_delivery_address():
    user = _get_current_user()
    if not user:
        return jsonify({"error": "User not found"}), 404

    data = request.get_json(silent=True) or {}
    try:
        address = create_delivery_address(user, data)
    except ValueError as exc:
        return jsonify({"error": str(exc)}), 400

    return jsonify({"address": address.to_dict()}), 201


@me_bp.route("/me/delivery-addresses/<address_id>", methods=["PATCH"])
@jwt_required()
def update_my_delivery_address(address_id: str):
    user = _get_current_user()
    if not user:
        return jsonify({"error": "User not found"}), 404

    try:
        aid = uuid.UUID(address_id)
    except ValueError:
        return jsonify({"error": "Invalid address ID"}), 400

    address = get_delivery_address(user, aid)
    if not address:
        return jsonify({"error": "Address not found"}), 404

    data = request.get_json(silent=True) or {}
    try:
        address = update_delivery_address(address, data)
    except ValueError as exc:
        return jsonify({"error": str(exc)}), 400

    return jsonify({"address": address.to_dict()})


@me_bp.route("/me/delivery-addresses/<address_id>", methods=["DELETE"])
@jwt_required()
def delete_my_delivery_address(address_id: str):
    user = _get_current_user()
    if not user:
        return jsonify({"error": "User not found"}), 404

    try:
        aid = uuid.UUID(address_id)
    except ValueError:
        return jsonify({"error": "Invalid address ID"}), 400

    address = get_delivery_address(user, aid)
    if not address:
        return jsonify({"error": "Address not found"}), 404

    delete_delivery_address(address)
    return jsonify({"message": "Address deleted"})


@me_bp.route("/me/delivery-addresses/<address_id>/set-default", methods=["POST"])
@jwt_required()
def set_my_default_delivery_address(address_id: str):
    user = _get_current_user()
    if not user:
        return jsonify({"error": "User not found"}), 404

    try:
        aid = uuid.UUID(address_id)
    except ValueError:
        return jsonify({"error": "Invalid address ID"}), 400

    address = get_delivery_address(user, aid)
    if not address:
        return jsonify({"error": "Address not found"}), 404

    address = set_default_delivery_address(address)
    return jsonify({"address": address.to_dict()})


@me_bp.route("/me/authorized-pickups", methods=["GET"])
@jwt_required()
def get_authorized_pickups():
    user = _get_current_user()
    if not user:
        return jsonify({"error": "User not found"}), 404

    pickups = list_authorized_pickups(user)
    return jsonify(
        {
            "pickups": [p.to_dict() for p in pickups],
            "max_pickups": MAX_AUTHORIZED_PICKUPS,
            "id_types": [
                {"value": t, "label": PICKUP_ID_TYPE_LABELS[t]} for t in PICKUP_ID_TYPES
            ],
        }
    )


@me_bp.route("/me/authorized-pickups", methods=["POST"])
@jwt_required()
def create_my_authorized_pickup():
    user = _get_current_user()
    if not user:
        return jsonify({"error": "User not found"}), 404

    data = request.get_json(silent=True) or {}
    try:
        pickup = create_authorized_pickup(user, data)
    except ValueError as exc:
        return jsonify({"error": str(exc)}), 400

    return jsonify({"pickup": pickup.to_dict()}), 201


@me_bp.route("/me/authorized-pickups/<pickup_id>", methods=["PATCH"])
@jwt_required()
def update_my_authorized_pickup(pickup_id: str):
    user = _get_current_user()
    if not user:
        return jsonify({"error": "User not found"}), 404

    try:
        pid = uuid.UUID(pickup_id)
    except ValueError:
        return jsonify({"error": "Invalid pickup ID"}), 400

    pickup = get_authorized_pickup(user, pid)
    if not pickup:
        return jsonify({"error": "Authorized pickup person not found"}), 404

    data = request.get_json(silent=True) or {}
    try:
        pickup = update_authorized_pickup(pickup, data)
    except ValueError as exc:
        return jsonify({"error": str(exc)}), 400

    return jsonify({"pickup": pickup.to_dict()})


@me_bp.route("/me/authorized-pickups/<pickup_id>", methods=["DELETE"])
@jwt_required()
def delete_my_authorized_pickup(pickup_id: str):
    user = _get_current_user()
    if not user:
        return jsonify({"error": "User not found"}), 404

    try:
        pid = uuid.UUID(pickup_id)
    except ValueError:
        return jsonify({"error": "Invalid pickup ID"}), 400

    pickup = get_authorized_pickup(user, pid)
    if not pickup:
        return jsonify({"error": "Authorized pickup person not found"}), 404

    delete_authorized_pickup(pickup)
    return jsonify({"message": "Authorized pickup person removed"})

File: server/src/routes/me.ts
import { Router } from 'express'
import {
  MAX_AUTHORIZED_PICKUPS,
  MAX_DELIVERY_ADDRESSES,
  PICKUP_ID_TYPE_LABELS,
  PICKUP_ID_TYPES,
} from '../constants.js'
import { authorizedPickupToDict } from '../lib/serializers/authorizedPickup.js'
import { deliveryAddressToDict } from '../lib/serializers/deliveryAddress.js'
import { userToDict } from '../lib/serializers/user.js'
import { jwtRequired, requireAuth, resolveJwtUser, type AuthRequest } from '../middleware/auth.js'
import {
  createAuthorizedPickup,
  deleteAuthorizedPickup,
  getAuthorizedPickup,
  listAuthorizedPickups,
  updateAuthorizedPickup,
} from '../services/authorizedPickupService.js'
import {
  createDeliveryAddress,
  deleteDeliveryAddress,
  getDeliveryAddress,
  listDeliveryAddresses,
  setDefaultDeliveryAddress,
  updateDeliveryAddress,
} from '../services/deliveryAddressService.js'
import { changePassword, updateProfile } from '../services/profileService.js'
import { routeParam } from '../lib/routeParams.js'
import { buildShippingAddress } from '../services/warehouseService.js'

export const meRouter = Router()

meRouter.get('/me', jwtRequired, requireAuth, (req: AuthRequest, res) => {
  const user = req.user!
  return res.json({
    user: userToDict(user, {
      includeTrn: user.role === 'customer',
      includeClerkFields: user.role === 'clerk' || user.role === 'admin',
    }),
  })
})

meRouter.patch('/me', jwtRequired, requireAuth, async (req: AuthRequest, res) => {
  try {
    const user = await updateProfile(req.user!, req.body ?? {})
    return res.json({ user: userToDict(user, { includeTrn: true }) })
  } catch (exc) {
    return res.status(400).json({ error: exc instanceof Error ? exc.message : String(exc) })
  }
})

meRouter.post('/me/change-password', jwtRequired, requireAuth, async (req: AuthRequest, res) => {
  const data = req.body ?? {}
  try {
    await changePassword(req.user!, String(data.current_password ?? ''), String(data.new_password ?? ''))
    return res.json({ message: 'Password updated' })
  } catch (exc) {
    return res.status(400).json({ error: exc instanceof Error ? exc.message : String(exc) })
  }
})

meRouter.get('/me/shipping-address', jwtRequired, requireAuth, (req: AuthRequest, res) => {
  const user = req.user!
  return res.json({
    shipping_id: user.shippingId,
    shipping_address: buildShippingAddress(user.shippingId),
  })
})

meRouter.get('/me/delivery-addresses', jwtRequired, requireAuth, async (req: AuthRequest, res) => {
  const addresses = await listDeliveryAddresses(req.user!)
  return res.json({
    addresses: addresses.map(deliveryAddressToDict),
    max_addresses: MAX_DELIVERY_ADDRESSES,
  })
})

meRouter.post('/me/delivery-addresses', jwtRequired, requireAuth, async (req: AuthRequest, res) => {
  try {
    const address = await createDeliveryAddress(req.user!, req.body ?? {})
    return res.status(201).json({ address: deliveryAddressToDict(address) })
  } catch (exc) {
    return res.status(400).json({ error: exc instanceof Error ? exc.message : String(exc) })
  }
})

meRouter.patch('/me/delivery-addresses/:addressId', jwtRequired, requireAuth, async (req: AuthRequest, res) => {
  const address = await getDeliveryAddress(req.user!, routeParam(req.params.addressId))
  if (!address) return res.status(404).json({ error: 'Address not found' })
  try {
    const updated = await updateDeliveryAddress(address, req.body ?? {})
    return res.json({ address: deliveryAddressToDict(updated) })
  } catch (exc) {
    return res.status(400).json({ error: exc instanceof Error ? exc.message : String(exc) })
  }
})

meRouter.delete('/me/delivery-addresses/:addressId', jwtRequired, requireAuth, async (req: AuthRequest, res) => {
  const address = await getDeliveryAddress(req.user!, routeParam(req.params.addressId))
  if (!address) return res.status(404).json({ error: 'Address not found' })
  await deleteDeliveryAddress(address)
  return res.json({ message: 'Address deleted' })
})

meRouter.post('/me/delivery-addresses/:addressId/set-default', jwtRequired, requireAuth, async (req: AuthRequest, res) => {
  const address = await getDeliveryAddress(req.user!, routeParam(req.params.addressId))
  if (!address) return res.status(404).json({ error: 'Address not found' })
  const updated = await setDefaultDeliveryAddress(address)
  return res.json({ address: deliveryAddressToDict(updated) })
})

meRouter.get('/me/authorized-pickups', jwtRequired, requireAuth, async (req: AuthRequest, res) => {
  const pickups = await listAuthorizedPickups(req.user!)
  return res.json({
    pickups: pickups.map(authorizedPickupToDict),
    max_pickups: MAX_AUTHORIZED_PICKUPS,
    id_types: PICKUP_ID_TYPES.map((t) => ({ value: t, label: PICKUP_ID_TYPE_LABELS[t] })),
  })
})

meRouter.post('/me/authorized-pickups', jwtRequired, requireAuth, async (req: AuthRequest, res) => {
  try {
    const pickup = await createAuthorizedPickup(req.user!, req.body ?? {})
    return res.status(201).json({ pickup: authorizedPickupToDict(pickup) })
  } catch (exc) {
    return res.status(400).json({ error: exc instanceof Error ? exc.message : String(exc) })
  }
})

meRouter.patch('/me/authorized-pickups/:pickupId', jwtRequired, requireAuth, async (req: AuthRequest, res) => {
  const pickup = await getAuthorizedPickup(req.user!, routeParam(req.params.pickupId))
  if (!pickup) return res.status(404).json({ error: 'Authorized pickup person not found' })
  try {
    const updated = await updateAuthorizedPickup(pickup, req.body ?? {})
    return res.json({ pickup: authorizedPickupToDict(updated) })
  } catch (exc) {
    return res.status(400).json({ error: exc instanceof Error ? exc.message : String(exc) })
  }
})

meRouter.delete('/me/authorized-pickups/:pickupId', jwtRequired, requireAuth, async (req: AuthRequest, res) => {
  const pickup = await getAuthorizedPickup(req.user!, routeParam(req.params.pickupId))
  if (!pickup) return res.status(404).json({ error: 'Authorized pickup person not found' })
  await deleteAuthorizedPickup(pickup)
  return res.json({ message: 'Authorized pickup person removed' })
})
